torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.cpp

```diff
@@ -17,171 +18,881 @@ namespace fuser {
+      ? ref->getRootDomain()
+      : ref->getMaybeRFactorDomain();
+  while (!root_dom.empty()) {
+    mappers.push_back(ContiguousInnerDimensionsMapper::map(ref, root_dom));
+    root_dom.erase(root_dom.begin());
+  }
+  return mappers;
+}
+
+// Returns Val* entires that should be evaluated and multiplied based on
+// contiguity of reference and dimensions mapped to ref in mapper.
+std::vector<Val*> getContigVectorSizesOf(
+    TensorView* of_tv,
+    const ContiguousInnerDimensionsMapper& mapper) {
+  // Logic copied to get root according to scheduler_utils::innerMostRootDim
+  // also copied from SchedulerRuntimeInfo::getMaxVectorizableWidth
+  bool use_root_dom = of_tv->hasReduction() && of_tv->hasRFactor();
+  auto of_tv_root =
+      use_root_dom ? of_tv->getRootDomain() : of_tv->getMaybeRFactorDomain();
+
+  const auto& projected_dim_map =
+      use_root_dom ? mapper.mappedRootIds() : mapper.mappedRFactorIds();
+
+  std::vector<IterDomain*> null_dims;
+  const std::vector<IterDomain*>& projected_dims =
+      projected_dim_map.find(of_tv) == projected_dim_map.end()
+      ? null_dims
+      : projected_dim_map.at(of_tv);
+
+  auto of_tv_root_no_reductions = TensorDomain::noReductions(of_tv_root);
+
+  auto contiguity = of_tv->domain()->contiguity();
+  // Appears after reductions the reduction domain often has a contiguity entry.
+  // This only matters if the result of the reduction is an output
+  if (contiguity.size() == of_tv_root.size() &&
+      contiguity.size() != of_tv_root_no_reductions.size()) {
+    std::vector<bool> new_contiguity;
+    for (auto i : c10::irange(of_tv_root.size())) {
+      if (!of_tv_root[i]->isReduction()) {
+        new_contiguity.push_back(contiguity[i]);
+      }
     }
-    vector_size = next_vector_size;
-    next_vector_size *= 2;
+    contiguity = new_contiguity;
   }
+  of_tv_root = of_tv_root_no_reductions;
 
-  return vector_size;
+  auto of_tv_root_size = of_tv_root.size();
+
+  // Filter out 0-dim tensors
+  if (of_tv_root_size < 1) {
+    return {};
+  }
+
+  // Filter out mismatched contiguity info
+  if (of_tv_root_size != contiguity.size()) {
+    return {};
+  }
+
+  std::vector<Val*> vectorizable_dim_sizes;
+
+  // Order is important, need to make sure dimensions match up correctly with
+  // what was propogated through the mapper. The mapper's dimensions is
+  // propogated in the order of the reference, if that order doesn't match the
+  // tensor we're mapping too then a transpose interfered with expanded the
+  // vectorize dimension.
+  size_t projected_dims_i = projected_dims.size();
+
+  for (auto i : c10::irange(of_tv_root_size)) {
+    if (projected_dims_i == 0) {
+      break;
+    }
+    auto root_i = of_tv_root_size - i - 1;
+    auto root_id = of_tv_root[root_i];
+
+    if (root_id->extent()->isOneInt() || root_id->isBroadcast()) {
+      if (projected_dims[projected_dims_i - 1]->sameAs(root_id)) {
+        --projected_dims_i;
+      }
+      continue;
+    }
```

Comment:
So we are ignoring the contiguity of broadcast dimensions? Today, if we have a tensor of shape `(4, 5, 1, 6)` and stride `(5000, 1000, 1000, 1)`, we will have contiguity `(true, true, false, true)`, this will pass the check here, but I don't think we can vectorize 4.

Reply:
Contiguity should be bounded by `getExpandedVectorization` function as the input `default_word_size` or the loop on `vectorizable_inputs_outputs`. Limitation for each tensor based on their contiguity should be taken into consideration, just not as part of this analysis.

Reply:
Note to self to double check this